presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java

```diff
@@ -217,59 +224,120 @@ public void close()
         afterClose.run();
     }
 
+    @VisibleForTesting
+    static NestedLoopOutputIterator createNestedLoopOutputIterator(Page probePage, Page buildPage)
+    {
+        if (probePage.getChannelCount() == 0 && buildPage.getChannelCount() == 0) {
+            int probePositions = probePage.getPositionCount();
+            int buildPositions = buildPage.getPositionCount();
+            try {
+                // positionCount is an int. Make sure the product can still fit in an int.
+                int outputPositions = multiplyExact(probePositions, buildPositions);
+                if (outputPositions <= PageProcessor.MAX_BATCH_SIZE) {
+                    return new PageRepeatingIterator(new Page(outputPositions), 1);
+                }
+            }
+            catch (ArithmeticException overflow) {
+            }
+            // Repeat larger position count a smaller position count number of times
+            Page outputPage = new Page(max(probePositions, buildPositions));
+            return new PageRepeatingIterator(outputPage, min(probePositions, buildPositions));
+        }
+        else if (probePage.getChannelCount() == 0 && probePage.getPositionCount() <= buildPage.getPositionCount()) {
+            return new PageRepeatingIterator(buildPage, probePage.getPositionCount());
+        }
+        else if (buildPage.getChannelCount() == 0 && buildPage.getPositionCount() <= probePage.getPositionCount()) {
+            return new PageRepeatingIterator(probePage, buildPage.getPositionCount());
+        }
+        else {
+            return new NestedLoopPageBuilder(probePage, buildPage);
+        }
+    }
+
+    // bi-morphic parent class for the two implementations allowed. Adding a third implementation will make getOutput megamorphic and
+    // should be avoided
+    @VisibleForTesting
+    abstract static class NestedLoopOutputIterator
```

Comment:
I think it's better to be an interface than an abstract static class since it's just describing what the implementors should do but do not share functional code.

Reply:
I thought about that and I'm still ok with changing this to an interface, but I initially preferred an abstract class with a comment about making the implementations bimorphic to make sure someone doesn't come along and refactor this into implementing `Iterator<Page>` and/or add another implementation to avoid making this megamorphic. See [the comments about this choice in prestosql's review](https://github.com/prestosql/presto/pull/5276#discussion_r499796050) for some more context. Let me know if you'd still prefer an `interface` there should be no performance difference after optimization by C2  since as long as the interface stays bimorphic.

Reply:
@pettyjamesm I don't think making it an Abstract class can prevent people making it megamorphic. That's why having a JMH benchmark covering these cases is important. Pasting before and after results would usually be very helpful and future regression, if any, can be easily identified. How did you test the performance otherwise? Was it something easy to merge to the code base? 
Other than this, the PR looks good. Thanks for making the changes.

 cc @mbasmanova

Reply:
Thanks. Performance was assessed by running targeted queries against a live cluster to test the empty channel scenarios as well as the "standard case". The standard case improvement is marginal but proportional to the number of columns since the allocation rate is the only major difference in the two scenarios. I don't currently have a JMH setup available so I'd have to implement that to put down some numbers.

Reply:
@pettyjamesm Thank you for letting me know the test results. JMH doesn't need any set up and can directly run in IntelliJ. Furthermore it can collect CPU profiles and GC profiles easily. I'll approve it for now but it'll be helpful to include JMH benchmark results for future performance improvement PRs. Thank you!

@mbasmanova Do you want to take another look at this PR?